### PR TITLE
PWX-24572 : Set and Get InstanceGroupSize implementation

### PR DESCRIPTION
**Pull request by nikita-bhatia**
Task : https://portworx.atlassian.net/browse/PWX-24572

Changes Done  : 
>  Added clusterId field in oracleOps struct
> SetInstanceGroupSize and GetInstanceGroupSize implementation

Tests done : 
> Upscaling and downscaling performed using cloudops tests 

image

### CODE REVIEW

File: oracle/oracle.go [hunk 1]
```diff
@@ -91,32 +96,37 @@ func NewClient() (cloudops.Ops, error) {
 
 func getInfoFromEnv(oracleOps *oracleOps) error {
 	var err error
-	oracleOps.instance, err = cloudops.GetEnvValueStrict(envInstanceID)
+	oracleOps.instance, err = cloudops.GetEnvValueStrict(envPrefix + "_" + envInstanceID)
```
Comment (vinayakshnd) [inline oracle/oracle.go:99]: envPrefix is not required. change it everywhere.

File: oracle/oracle.go [hunk 2]
```diff
@@ -444,3 +458,125 @@ func (o *oracleOps) Delete(volumeID string) error {
...
+
+func (o *oracleOps) SetInstanceGroupSize(instanceGroupID string, count int64, timeout time.Duration) error {
+
+	ctx := context.Background()
+	var cancel context.CancelFunc
+	if timeout > time.Nanosecond {
+		// If timeout is non-zero then
+		// add deadline to the operation else
+		// wait infinitely for operation to complete
+		ctx, cancel = context.WithTimeout(ctx, timeout)
```
Comment (vinayakshnd) [inline oracle/oracle.go:470]: this `ctx` object is not used anywhere (in ListNodePools or in UpdateNodePool request)

File: oracle/oracle.go [hunk 3]
```diff
@@ -444,3 +458,125 @@ func (o *oracleOps) Delete(volumeID string) error {
...
+		defer cancel()
+	}
+
+	instanceGroupSize := int(count)
+
+	//get containerengine client
+	client, err := containerengine.NewContainerEngineClientWithConfigurationProvider(common.DefaultConfigProvider())
```
Comment (vinayakshnd) [inline oracle/oracle.go:477]: `oracleOps` already has containerEngine client. We can use that. No need to initialize it here again.

File: oracle/oracle.go [hunk 4]
```diff
@@ -444,3 +458,125 @@ func (o *oracleOps) Delete(volumeID string) error {
...
+		nodePoolPlacementConfigDetails[i].SubnetId = placementConfigs.SubnetId
+	}
+
+	//update node pools
+	req := containerengine.UpdateNodePoolRequest{
+		NodePoolId: nodePools.Items[0].Id, //get node pool id
+		UpdateNodePoolDetails: containerengine.UpdateNodePoolDetails{
+			NodeConfigDetails: &containerengine.UpdateNodePoolNodeConfigDetails{
+				Size:             &totalClusterSize,
+				PlacementConfigs: nodePoolPlacementConfigDetails,
```
Comment (vinayakshnd) [inline oracle/oracle.go:510]: Here, we are not making any changes to placementConfig and it's not a required (must) field. We can skip creating a duplicate `nodePoolPlacementConfigDetails`, if oracle SDK allows that.

File: oracle/oracle.go [hunk 5]
```diff
@@ -444,3 +458,125 @@ func (o *oracleOps) Delete(volumeID string) error {
...
+			},
+		},
+	}
+
+	resp, err := client.UpdateNodePool(context.Background(), req)
+	if err != nil {
+		return err
+	}
+
+	err = o.waitNodeStatus(resp.OpcRequestId, resp.OpcWorkRequestId, client)
```
Comment (vinayakshnd) [inline oracle/oracle.go:520]: Rename function `waitNodeStatus` to `waitNodePoolStatus`

File: oracle/oracle.go [hunk 6]
```diff
@@ -444,3 +458,125 @@ func (o *oracleOps) Delete(volumeID string) error {
...
+
+	return nil
+}
+
+func (o *oracleOps) waitNodeStatus(opcRequestID, opcWorkRequestID *string, client containerengine.ContainerEngineClient) error {
+	workReq := containerengine.GetWorkRequestRequest{OpcRequestId: opcRequestID,
+		WorkRequestId: opcWorkRequestID}
+
+	f := func() (interface{}, bool, error) {
+		workResp, err := client.GetWorkRequest(context.Background(), workReq)
```
Comment (vinayakshnd) [inline oracle/oracle.go:533]: [GetNodePool](https://docs.oracle.com/en-us/iaas/api/#/en/containerengine/20180222/NodePool/GetNodePool) API needs to be used to check lifecycle state of node pool is `ACTIVE`.

As additional check, (if required,) call `GetInstanceGroupSize` to verify number of active worker nodes matches with expected count.

File: oracle/oracle.go [hunk 7]
```diff
@@ -444,3 +458,125 @@ func (o *oracleOps) Delete(volumeID string) error {
...
+
+		logrus.Debugf("Node status is in [%s] state", workResp.Status)
+		return nil, true, fmt.Errorf("Node status is still in [%s] state", workResp.Status)
+	}
+	_, err := task.DoRetryWithTimeout(f, 5*time.Minute, 10*time.Second)
+	return err
+}
+
+func (o *oracleOps) GetInstanceGroupSize(instanceGroupID string) (int64, error) {
+	client, err := containerengine.NewContainerEngineClientWithConfigurationProvider(common.DefaultConfigProvider())
```
Comment (vinayakshnd) [inline oracle/oracle.go:550]: use `oracleOps. containerEngine` instead of re-initialzing here.

File: oracle/oracle.go [hunk 8]
```diff
@@ -444,3 +458,125 @@ func (o *oracleOps) Delete(volumeID string) error {
...
+
+	var count int64
+
+	nodePoolReq := containerengine.ListNodePoolsRequest{CompartmentId: &o.compartmentID, Name: &instanceGroupID, ClusterId: &o.clusterID}
+	nodePools, err := client.ListNodePools(context.Background(), nodePoolReq)
+	if err != nil {
+		return 0, err
+	}
+
+	req := containerengine.GetNodePoolRequest{NodePoolId: nodePools.Items[0].Id}
```
Comment (vinayakshnd) [inline oracle/oracle.go:563]: Perform check for number of items in `nodePools.Items` array before this function call.

**Review by vinayakshnd — CHANGES_REQUESTED** [review]

File: oracle/oracle.go [hunk 9]
```diff
@@ -444,3 +458,125 @@ func (o *oracleOps) Delete(volumeID string) error {
...
+
+	err = o.waitNodeStatus(resp.OpcRequestId, resp.OpcWorkRequestId, client)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (o *oracleOps) waitNodeStatus(opcRequestID, opcWorkRequestID *string, client containerengine.ContainerEngineClient) error {
```
Comment (vinayakshnd) [inline oracle/oracle.go:528]: No need to pass `client` as input argument. Instead take desired state as input argument.

File: oracle/oracle.go (hunk 4 above)
Comment (nikita-bhatia) [reply oracle/oracle.go:498]: nodePoolPlacementConfigDetails is required to equally balance creation and deletion of instances in each availability domain. We need to provide one placement configuration for each availability domain in which we intend to launch a node.

File: oracle/oracle.go (hunk 8 above)
Comment (nikita-bhatia) [reply oracle/oracle.go:551]: Done

File: oracle/oracle.go (hunk 2 above)
Comment (nikita-bhatia) [reply oracle/oracle.go]: Removed

File: oracle/oracle.go (hunk 3 above)
Comment (nikita-bhatia) [reply oracle/oracle.go]: Done

File: oracle/oracle.go (hunk 5 above)
Comment (nikita-bhatia) [reply oracle/oracle.go]: Changed to waitTillWorkStatusIsSucceeded to make it generic for any work request.

File: oracle/oracle.go (hunk 6 above)
Comment (nikita-bhatia) [reply oracle/oracle.go]: The state of work request changes from Accepted-> InProgress -> Succeeded, till all the instances are active. Oracle SDK provides this functionality to track the status  of any work request submitted till done.

File: oracle/oracle.go (hunk 7 above)
Comment (nikita-bhatia) [reply oracle/oracle.go]: done

File: oracle/oracle.go (hunk 9 above)
Comment (nikita-bhatia) [reply oracle/oracle.go]: done

File: oracle/oracle.go (hunk 1 above)
Comment (nikita-bhatia) [reply oracle/oracle.go]: Done

**Review by vinayakshnd — APPROVED** [review]
LGTM

**Pull request merged**